setup/puppeteer_environment.js

```diff
@@ -0,0 +1,40 @@
+// puppeteer_environment.js
+const print = require('./print');
+const NodeEnvironment = require('jest-environment-node');
+const fs = require('fs');
+const path = require('path');
+const puppeteer = require('puppeteer');
+const os = require('os');
+const DIR = path.join(os.tmpdir(), 'jest_puppeteer_global_setup');
+
+class PuppeteerEnvironment extends NodeEnvironment {
+  constructor(config) {
+    super(config);
+  }
+
+  async setup() {
+    print('Setup Test Environment.',"warning")
+    await super.setup();
+    // get the wsEndpoint
+    const wsEndpoint = fs.readFileSync(path.join(DIR, 'wsEndpoint'), 'utf8');
+    if (!wsEndpoint) {
+      throw new Error('wsEndpoint not found');
+    }
+
+    // connect to puppeteer
+    this.global.__BROWSER__ = await puppeteer.connect({
+      browserWSEndpoint: wsEndpoint,
+    });
+  }
+
+  async teardown() {
```

Comment:
There are teardown function and teardown.js file. What is the difference?

Reply:
We can close the **puppeteer instance** and clean-up the file.
This is only for one custom puppeteer instance's teardown.

For example imagine 1 window chrome browser is Global Browser < which is created with setup.js
  - Every tab is new page for puppeteer created from puppeteer_environment.js .
  - firstly we are init global setup than each test puppeteer .
  - than closing it with this teardown function.
  - Finally after complete each test , We are closing global browser instance with teardown.js 

For example , We can create different puppeteer instance for each different test file.
**But teardown.js is closing global browser.**